spring-cloud-dataflow-module-deployer-spi/src/main/java/org/springframework/cloud/dataflow/module/deployer/test/AbstractModuleDeployerTests.java

```diff
@@ -0,0 +1,166 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.module.deployer.test;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.springframework.cloud.dataflow.module.DeploymentState.deployed;
+import static org.springframework.cloud.dataflow.module.DeploymentState.unknown;
+import static org.springframework.cloud.dataflow.module.deployer.test.EventuallyMatcher.eventually;
+
+import java.util.UUID;
+
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
+import org.hamcrest.collection.IsMapContaining;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.dataflow.core.ArtifactCoordinates;
+import org.springframework.cloud.dataflow.core.ModuleDefinition;
+import org.springframework.cloud.dataflow.core.ModuleDeploymentId;
+import org.springframework.cloud.dataflow.core.ModuleDeploymentRequest;
+import org.springframework.cloud.dataflow.module.ModuleStatus;
+import org.springframework.cloud.dataflow.module.deployer.ModuleDeployer;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ * Abstract base class for integration tests of
+ * {@link org.springframework.cloud.dataflow.module.deployer.ModuleDeployer} implementations.
+ *
+ * <p>Inheritors should setup an environment with a newly created {@link ModuleDeployer} that has no pre-deployed
+ * modules. Tests in this class are independent and leave the deployer in a clean state after they successfully run.</p>
+ *
+ * <p>As deploying a module is often quite time consuming, some tests often test various aspects of deployment in a
+ * row, to avoid re-deploying modules over and over again.</p>
+ *
+ * @author Eric Bottard
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+public abstract class AbstractModuleDeployerTests {
+
+	@Autowired
+	protected ModuleDeployer moduleDeployer;
+
+	@Test
+	public void newlyCreatedDeployerShouldReportNoModules() {
+		assertThat(moduleDeployer.status().isEmpty(), is(true));
+	}
+
+	@Test
+	public void testUnknownDeployment() {
+		ModuleDeploymentId id = new ModuleDeploymentId("agroup", "alabel");
+		ModuleStatus status = moduleDeployer.status(id);
+
+		assertThat(status.getModuleDeploymentId(), is(id));
+		assertThat("The map was not empty: " + status.getInstances(), status.getInstances().isEmpty(), is(true));
+		assertThat(status.getState(), is(unknown));
+	}
+
+	@Test
+	public void testSimpleDeployment() {

```

Comment:
I know this is WIP and quite early, so this is a improvement suggestion - but a 'module redeployment' core test here would be helpful, too.